engine/kv_engine_hash.cpp

```diff
@@ -0,0 +1,242 @@
+  if (s != Status::Ok) {
+    return s;
+  }
+  auto token = version_controller_.GetLocalSnapshotHolder();
+  *len = hlist->Size();
+  return Status::Ok;
+}
+
+Status KVEngine::HashGet(StringView key, StringView field, std::string* value) {
+  if (!CheckKeySize(key) || !CheckKeySize(field)) {
+    return Status::InvalidDataSize;
+  }
+
+  HashList* hlist;
+  Status s = hashListFind(key, &hlist, false);
+  if (s != Status::Ok) {
+    return s;
+  }
+  auto token = version_controller_.GetLocalSnapshotHolder();
+  LookupResult result =
+      lookupImpl<false>(hlist->InternalKey(field), RecordType::HashElem);
+  if (result.s != Status::Ok) {
+    return result.s;
+  }
+  StringView val = result.entry.GetIndex().dl_record->Value();
+  value->assign(val.data(), val.size());
+  return Status::Ok;
+}
+
+Status KVEngine::HashSet(StringView key, StringView field, StringView value) {
+  if (!CheckKeySize(key) || !CheckKeySize(field) || !CheckValueSize(value)) {
+    return Status::InvalidDataSize;
+  }
+
+  HashList* hlist;
+  Status s = hashListFind(key, &hlist, true);
+  if (s != Status::Ok) {
+    return s;
+  }
+
+  auto internal_key = hlist->InternalKey(field);
+  auto guard = hash_table_->AcquireLock(internal_key);
+  LookupResult result = lookupImpl<true>(internal_key, RecordType::HashElem);
+  if (!(result.s == Status::Ok || result.s == Status::NotFound)) {
+    return result.s;
+  }
+
+  auto token = version_controller_.GetLocalSnapshotHolder();
+  TimeStampType ts = token.Timestamp();
+  auto space = pmem_allocator_->Allocate(sizeof(DLRecord) +
+                                         internal_key.size() + value.size());
+  if (space.size == 0) {
+    return Status::PmemOverflow;
+  }
+  void* addr = pmem_allocator_->offset2addr_checked(space.offset);
+  if (result.s == Status::NotFound) {
+    if (std::hash<StringView>{}(field) % 2 == 0) {
+      hlist->PushFrontWithLock(space, ts, field, value);
+    } else {
+      hlist->PushBackWithLock(space, ts, field, value);
+    }
+    insertImpl(result, internal_key, RecordType::HashElem, addr);
+  } else {
+    DLRecord* old_rec = result.entry.GetIndex().dl_record;
+    hlist->ReplaceWithLock(space, old_rec, ts, field, value,
+                           [&](DLRecord* rec) { delayFree(rec, ts); });
+    insertImpl(result, internal_key, RecordType::HashElem, addr);
+  }
+  return Status::Ok;
+}
+
+Status KVEngine::HashDelete(StringView key, StringView field) {
+  if (!CheckKeySize(key) || !CheckKeySize(field)) {
+    return Status::InvalidDataSize;
+  }
+  HashList* hlist;
+  Status s = hashListFind(key, &hlist, false);
+  if (s == Status::NotFound) {
+    return Status::Ok;
+  }
+  if (s != Status::Ok) {
+    return s;
+  }
+
+  LookupResult ret;
+  {
+    auto guard = hash_table_->AcquireLock(hlist->InternalKey(field));
+    ret = removeImpl(hlist->InternalKey(field), RecordType::HashElem);
+  }
+  if (ret.s == Status::NotFound) {
+    return Status::Ok;
+  }
+  if (ret.s != Status::Ok) {
```

Comment:
if `Status::OutDated`, need to update entry_status to `OutDated`.  Only check hash entry status to avoid read pmem once during scanning table. please align to other place.